pandas/core/frame.py

```diff
@@ -7575,6 +7575,19 @@ def _count_level(self, level, axis=0, numeric_only=False):
     def _reduce(
         self, op, name, axis=0, skipna=True, numeric_only=None, filter_type=None, **kwds
     ):
+
+        dtype_is_dt = self.dtypes.apply(lambda x: x.kind == "M")
+        if numeric_only is None and name in ["mean", "median"] and dtype_is_dt.any():
+            warnings.warn(
```

Comment:
Users will need an option to actually do anything about this now, and silence the warning.

IIUC, we would need something like `datetime_is_numeric=None/True/False`, that's available to the user (in DataFrame.median) and passed through to here. And we would only warn for `datetime_is_numeric=None`.

Reply:
id like to move away from having a `numeric_only=None` case or others like it.  my preference would be to have users do something like `df.select_dtypes(exclude="datetime").mean()`

Reply:
I may have misread earlier. `numeric_only=False` will include datetime columns in these reductions, without a warning? So we should be OK here.

> id like to move away from having a numeric_only=None case or others like it.

I share that goal.

Reply:
ok with the warning, we would need a deprecation cycle on numeric_only itself (ok as separate PR).

Reply:
ideally can we move this code (and to be honest _reduce *stuff* to nanops / ops) very messy here.

Reply:
the cleanup of _reduce that im looking forward to comes after #29847.

Reply:
k